### ISSUE TITLE
Experimental OptimizeCss flag (critters) isn't deferring all css files properly.

### ISSUE THREAD

**Issue by mlstubblefield** [vercel/next.js#29925]
### What version of Next.js are you using?

11.1.3-canary.0

### What version of Node.js are you using?

16.1.0

### What browser are you using?

chrome (shouldn't matter, the html is what's different)

### What operating system are you using?

Windows (I think this could have to do with it)

### How are you deploying your application?

Vercel!

### Describe the Bug

I'm experimenting by using the new optmizeCss experimental property in the next.config.js (and it's sorta awesome for FCP).
Locally, I was able to get my FCP to be the same as TTFB. Hats off to you all!

So locally it works great, but when I deploy to vercel I don't have the same thing happen. Unfortunately, several of my CSS resources aren't properly defered and end up being render blocking (in a way that's worse than having the optimizeCss flag off).

Locally (npm run build/npm start) (next build/next start), the output HTML of my application has the defers CSS loading properly like so...

```
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/c6011bfa7bebf07f295b.css" data-n-g="" media="print"
  onload="this.media='all'"><noscript>
  <link rel="stylesheet" href="/mfe/pdp/_next/static/css/c6011bfa7bebf07f295b.css">
</noscript>
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/34299f6482a8336ff401.css" data-n-p="" media="print"
  onload="this.media='all'"><noscript>
  <link rel="stylesheet" href="/mfe/pdp/_next/static/css/34299f6482a8336ff401.css">
</noscript>
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/4f17191406a61c110eda.css" media="print"
  onload="this.media='all'"><noscript>
  <link rel="stylesheet" href="/mfe/pdp/_next/static/css/4f17191406a61c110eda.css">
</noscript>
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/0c5bcc4e2145116a961a.css" media="print"
  onload="this.media='all'"><noscript>
  <link rel="stylesheet" href="/mfe/pdp/_next/static/css/0c5bcc4e2145116a961a.css">
</noscript>
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/ba51bed5c56d6db12683.css" media="print"
  onload="this.media='all'"><noscript>
  <link rel="stylesheet" href="/mfe/pdp/_next/static/css/ba51bed5c56d6db12683.css">
</noscript>
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/968ed11759396a8e89e7.css" media="print"
  onload="this.media='all'"><noscript>
  <link rel="stylesheet" href="/mfe/pdp/_next/static/css/968ed11759396a8e89e7.css">
</noscript>
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/dbf9f1ea1a43e865f4fa.css" media="print"
  onload="this.media='all'"><noscript>
  <link rel="stylesheet" href="/mfe/pdp/_next/static/css/dbf9f1ea1a43e865f4fa.css">
</noscript>
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/90228b002c6938abeaf2.css" media="print"
  onload="this.media='all'"><noscript>
  <link rel="stylesheet" href="/mfe/pdp/_next/static/css/90228b002c6938abeaf2.css">
</noscript><noscript data-n-css=""></noscript>
```

On Vercel, I'm seeing this (some files deferred where as some are not)
```
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/c6011bfa7bebf07f295b.css" data-n-g="" media="print"
  onload="this.media='all'"><noscript>
  <link rel="stylesheet" href="/mfe/pdp/_next/static/css/c6011bfa7bebf07f295b.css">
</noscript>
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/34299f6482a8336ff401.css" data-n-p="" media="print"
  onload="this.media='all'"><noscript>
  <link rel="stylesheet" href="/mfe/pdp/_next/static/css/34299f6482a8336ff401.css">
</noscript>
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/4f17191406a61c110eda.css">
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/0c5bcc4e2145116a961a.css">
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/ba51bed5c56d6db12683.css">
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/968ed11759396a8e89e7.css">
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/dbf9f1ea1a43e865f4fa.css">
<link rel="stylesheet" href="/mfe/pdp/_next/static/css/90228b002c6938abeaf2.css">
```

You can see the last bunch of stylesheets starting with 4f17191406a61c110eda.css aren't properly deferred.

### Expected Behavior

The output HTML document from vercel would match the one on localhost.
IE: the deferred link tags (which work locally) would be the same when hosted on vercel.

### To Reproduce

I'm not able to get a minimal repro going for this unfortunately as I don't know quite what is triggering things.
I'm happy to communicate about the characteristics of our repository/set up that might be leading to this.

A couple of notable things... 

- We use styled components, however I'm quite sure that we're always using internal CSS for those
- We use next/dynamic quite a bit
- The files that are not being deferred are all css (sass) modules (we use the sass npm module), but that's also true for the ones that are deferred
- Some of the styles come from an npm package (a design library) that is being transpiled using next-transpile-modules, but the files that do/don't have the issue are a mixture (similar to the above point)

Here's a network capture... you'll notice that the size doesn't seem to be a factor (of course I don't know too much about the internals of critters)
![image](https://user-images.githubusercontent.com/7562533/137395573-be499b28-d4b4-42bb-974f-dbdbf18e8f60.png)

**Comment by awareness481** [comment]
> After a little more research, I noticed that the sheets that have the problem (are not deferred) are ironically in components that are loaded via next/dynamic.

Since the components are loaded dynamically I think the assumption is made that the styles do not need to deferred.

However, it can probably be considered a bug for dynamic components that are loaded immediately but are not on the initial viewport? :thinking:

**Comment by kamal-choudhary** [comment]
Hi @mlstubblefield, I'm also trying to implement this experimental optimizeCss in my project.

And I'm also using the Styled Components library.

But I'm confused that whether we actually need to implement this.

Because on the Styled Components docs page, it is written that critical CSS is handled automatically.

https://styled-components.com/docs/basics

P.S: I'm sorry if this is a noob question.

**Comment by mlstubblefield** [comment]
For styled components, you probably wouldn't (it kinda depends). The long short is, if you a link tag for a stylesheet in your CSS, then this might be beneficial.

**Comment by kamal-choudhary** [comment]
I got it. Matt, thank you so much for taking time to answer my question!

**Comment by janicklas-ralph** [comment]
Hi @mlstubblefield. If the styles from next/dynamic aren't present in the SSR response then they aren't evaluated for critical css. This usually works well since next/dynamic styles are added after hydration and critical css helps way before that during FCP.

Let me know if they CSS are present in the SSR response or not?

**Comment by mlstubblefield** [comment]
The styles are in the SSR response. I'm viewing the raw HTML server output before any JS runs on it. The example HTML in my first example is the raw HTML response from the server.
It doesn't make sense to me. It seems like critters should handle that perfectly.

**Comment by mlstubblefield** [comment]
Ah ok, I'll give that a try and post my results here.

**Comment by mlstubblefield** [comment]
Anything I can try here?

**Comment by mlstubblefield** [comment]
Any updates?

**Comment by mlstubblefield** [comment]
I lean towards this being more of a vercel bug VS something from the google team. It works when I do a prod build on my machine, it just breaks when I deploy to vercel.

**Comment by mlstubblefield** [comment]
Yea it works on my machine with "npm run build and npm run start", but not when deployed to vercel.

**Comment by mlstubblefield** [comment]
I found the issue-- in the NPM project that we were pulling in, we had things set up so that we could pull the project into a next app as well as a non next app.
In order to do so, we used loadable components and aliases....

TLDR changing our NPM package code from
```const EmailSignUpModalContent = loadable(() => import('./EmailSignUpModalContent'))```
to
```const EmailSignUpModalContent = dynamic(/* #__LOADABLE__ */() => import('./EmailSignUpModalContent'))```
seems to have addressed the issue. It appears that both loadable and next dynamic do some similar things with respect to babel transforms.

**Comment by mlstubblefield** [comment]
I think I made a mistake with my above post. It appears that in my vercel environment things are wrong again. It has me wondering if I made a mistake while evaluating my results.

**Comment by mlstubblefield** [comment]
Aha! I have another clue...

It works on one route of my site hosted in vercel "/", but doesn't work on a nother one "/*,*.html".
Something worth mentioning-- we're using assetPrefix to do some unique things with files.
So I think maybe it's either asset prefix or the custom routing that vercel is having trouble with.

**Comment by mlstubblefield** [comment]
Yea it was/is that-- we use a custom route. It works on the "regular" (next style) route, but not on the custom one.

**Comment by mlstubblefield** [comment]
Oh wow, I am taking crazy pills-- after I refreshed it went away. Let me switch environment sand reproduce that...
Before the refresh
![image](https://user-images.githubusercontent.com/7562533/150882427-842c5306-62cc-4740-aca4-dfe98f71a08f.png)
After the refresh
![image](https://user-images.githubusercontent.com/7562533/150882471-d81f0f0d-e179-4690-8292-f5f847a4dea6.png)

**Comment by mlstubblefield** [comment]
I have a bit of a guess that the way critters applies to a static page is slightly different from a non-static page...
Is it possible that critter is trying to run twice and causing havoc?
That'd cause it to work the first time, and then break once it's cached.
https://github.com/vercel/next.js/blob/41614e52d72f532601dc2d118f4f40dbd1e63907/packages/next/server/render.tsx#L1481

**Comment by mlstubblefield** [comment]
I can't for the life of me get it to be fixed once again. Another theory that I have is mentionings of "moving to new infrastructure". Perhaps a split test for moving to that could have fixed this intermittently?

**Comment by mlstubblefield** [comment]
More clues...

It always works for a page that has no parameters -- even if the headers say it's a cache miss.
If I redeploy it works when I get the header "x-vercel-cache: PRERENDER"
It does not work if the page responds with "x-vercel-cache: HIT"
I made a SSR'd version of the page (exports getServerSideProps) that always send back a cache miss. It never works. (there's a chance it did the first time).

**Comment by mlstubblefield** [comment]
I've got a hunch that I'm going to experiment with-- the 2 files that work are a decent amount larger than the others.
![image](https://user-images.githubusercontent.com/7562533/150890133-766215ba-f5b9-48f2-a170-c3c0520c3c8e.png)

**Comment by ijjk** [comment]
Hi, this has been updated in `v12.0.10-canary.0` of Next.js, please update and give it a try!

**Comment by mlstubblefield** [comment]
It works-- and we were able to improve our FCP pretty significantly as a result! Thanks!
![image](https://user-images.githubusercontent.com/7562533/152440050-46b658a8-5c43-4f11-9897-4cac8617e149.png)
